File: Problem1/haskell/multiples_of_3_and_5.hs
sum_multiples_of_3_and_5_below :: Integer -> Integer
sum_multiples_of_3_and_5_below num = sum(unique_multiples)
  where
    multiples = find_multiples_of_3_and_5_below num
    unique_multiples = remove_duplicates(multiples)

find_multiples_of_3_and_5_below :: Integer -> [Integer]
find_multiples_of_3_and_5_below num = multiples_of_3 ++ multiples_of_5
  where
    multiples_of_3 = find_multiples_of_3_below num 
    multiples_of_5 = find_multiples_of_5_below num 

remove_duplicates :: [Integer] -> [Integer]
remove_duplicates [] = []
remove_duplicates(x:xs) = remove_duplicates ys ++ [x] ++ remove_duplicates zs
  where
    ys = [a | a <- xs, a < x]
    zs = [b | b <- xs, b > x]

find_multiples_of_3_below :: Integer -> [Integer]
find_multiples_of_3_below num = [x | x <- [1..(num-1)], x `mod` 3 == 0]

find_multiples_of_5_below :: Integer -> [Integer]
find_multiples_of_5_below num = [x | x <- [1..(num-1)], x `mod` 5 == 0]

File: Problem1/ruby/lib/multiples_of_3_and_5.rb
class MultiplesOf3And5
  def sum_of_multiples_of_3_and_5_below(x)
    numbers = numbers_below(x)
    multiples = multiples_of_3(numbers) + multiples_of_5(numbers)
    unique_multiples = multiples.uniq
    unique_multiples.inject(:+)
  end
  
  def numbers_below(max)
    (1..max - 1).to_a
  end

  def multiples_of_3(list)
    list.select do |n|
      n % 3 == 0
    end
  end

  def multiples_of_5(list)
    list.select do |n|
      n % 5 == 0
    end
  end
end
